test/sql/insert_sql_test.cpp

```diff
@@ -580,6 +620,76 @@ TEST_F(InsertSQLTests, NonExistentTable) {
   txn = txn_manager.BeginTransaction();
   EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
                peloton::CatalogException);
+
+  // free the database just created
+  txn = txn_manager.BeginTransaction();
+  catalog::Catalog::GetInstance()->DropDatabaseWithName(DEFAULT_DB_NAME, txn);
+  txn_manager.CommitTransaction(txn);
+}
+
+TEST_F(InsertSQLTests, BadInserts) {
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+  catalog::Catalog::GetInstance()->CreateDatabase(DEFAULT_DB_NAME, txn);
+  txn_manager.CommitTransaction(txn);
+
+  CreateAndLoadTable8();
+
+  std::vector<ResultValue> result;
+  std::vector<FieldInfo> tuple_descriptor;
+  std::string error_message;
+  int rows_changed;
+  std::unique_ptr<optimizer::AbstractOptimizer> optimizer(
+      new optimizer::Optimizer());
+  rows_changed = 0;
+  // Insert a tuple with more values than target columns
+  std::string query("INSERT INTO test8 VALUES(1, 2, 3, 4);");
+  txn = txn_manager.BeginTransaction();
+  EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
+               peloton::CatalogException);
+  EXPECT_EQ(0,
+            rows_changed);  // no rows should be affected by the incorrect query
+
+  // Insert a tuple with more target columns than values
+  query = "INSERT INTO test8(num1, num3) VALUES(3);";
+  txn = txn_manager.BeginTransaction();
+  EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
+               peloton::CatalogException);
+  EXPECT_EQ(0,
+            rows_changed);  // no rows should be affected by the incorrect query
+
+  // Insert a tuple with more target columns than values
+  query = "INSERT INTO test8(num1, num3) VALUES (1, 2), (3, 4), (3, 4, 5);";
+  txn = txn_manager.BeginTransaction();
+  EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
+               peloton::CatalogException);
+  EXPECT_EQ(0, rows_changed);
+
+  // Insert a tuple with more target columns than values
+  query = "INSERT INTO test8(num1, num3) VALUES (6, 7), (3, 4, 5);";
+  txn = txn_manager.BeginTransaction();
+  EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
+               peloton::CatalogException);
+  EXPECT_EQ(0, rows_changed);
+
+  // Insert a tuple with more target columns than values
+  query = "INSERT INTO test8(numx) VALUES(3);";
+  txn = txn_manager.BeginTransaction();
+  EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
+               peloton::CatalogException);
+  EXPECT_EQ(0, rows_changed);
+
+  // Insert a tuple with more target columns than values
+  query = "INSERT INTO test8(num1, num4) VALUES(3, 4);";
+  txn = txn_manager.BeginTransaction();
+  EXPECT_THROW(TestingSQLUtil::GeneratePlanWithOptimizer(optimizer, query, txn),
+               peloton::CatalogException);
+  EXPECT_EQ(0, rows_changed);
```

Comment:
Be sure to add tests cases to check each of the four code paths you've added. I think you're missing the case where target list is implicit and number of expressions is less than the number of columns.

Reply:
@pmenon There's an underlying bug in that test case. I built the master branch and then tried to insert a single value in a table with two attributes - and that resulted in segmentation fault. I'll try to fix it in this PR, or should I open another PR for that?

Reply:
@pmenon I've added that test case and the additional code to handle it.